client/src/components/RatingsAndReviews/AddReviewForm.jsx

```diff
@@ -0,0 +1,418 @@
+/* eslint-disable no-param-reassign */
+import React from 'react';
+import PropTypes from 'prop-types';
+import axios from 'axios';
+import AddCharacteristic from './AddCharacteristic';
+
+const AddReviewForm = class extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      chars: [],
+      showStarLabel: false,
+      showRatingWarning: false,
+      innerWidth: '0%',
+      rating: null,
+      summary: '',
+      body: '',
+      recommend: false,
+      name: '',
+      email: '',
+      photos: [],
+      characteristics: {},
+    };
+
+    this.processCharacteristics = this.processCharacteristics.bind(this);
+    this.handleStarRatingClick = this.handleStarRatingClick.bind(this);
+    this.handleCharRatingClick = this.handleCharRatingClick.bind(this);
+    this.handleInputChange = this.handleInputChange.bind(this);
+    this.fileInput = React.createRef();
+    this.handleFileUpload = this.handleFileUpload.bind(this);
+    this.handleFormSubmit = this.handleFormSubmit.bind(this);
+  }
+
+  componentDidMount() {
+    this.processCharacteristics();
+  }
+
+  componentDidUpdate(prevProps) {
+    const { characteristics } = this.props;
+    if (JSON.stringify(characteristics) !== JSON.stringify(prevProps.characteristics)) {
```

Comment:
ah! Interesting, JSON.stringify to make sure they indeed not equivalent. But be careful with it! JS doesn't guarantee that the order of they keys will be the same, so that could leave the line open for a potential bug. 
https://stackoverflow.com/questions/15376185/is-it-fine-to-use-json-stringify-for-deep-comparisons-and-cloning

Reply:
Agreed, we can improve this with a helper library (underscore or lodash) to make sure this comparison of two objects is accurate. JSON.stringify will trigger things more often than not